sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1022 @@
+        atol = 1e-4
+        sample_map = find_ref_idxs(times, self.times)
+        mapped = np.abs(times[sample_map] - self.times) < atol
+
+        oframe_idxs = frame_idxs[sample_map]  # out-frame idx for each in sample
+        oframe_idxs[~mapped] = -1 
+        _, offsets = np.unique(frame_idxs, return_index=True)
+        # Sample idx within each out-frame for every input sample
+        out_offset_idxs = sample_map - offsets[frame_idxs[sample_map]]
+
+        iframe_idxs = self.frame_idxs  #in-frame idx for each in sample
+        _, offsets = np.unique(self.frame_idxs, return_index=True)
+        # Sample idx within each in-frame for every input sample
+        in_offset_idxs = np.arange(len(self.times)) - offsets[self.frame_idxs]
+
+        # Handle file writers
+        writer = None
+        cur_file_idx = None
+        out_files = []
+
+        inframe_iter = get_frame_iter(self.files)
+        iframe, interm_frames = next_scan(inframe_iter)
+        iframe_idx = 0
+        oframe_num = 0
+        pbar.update()
+
+        for oframe_idx in np.unique(frame_idxs):
+            # Update writer
+            if file_idxs[oframe_idx] != cur_file_idx:
+                close_writer(writer)
+                cur_file_idx = file_idxs[oframe_idx]
+                fname = os.path.join(
+                    outdir, f'D_{self.stream_id}_{cur_file_idx:0>3}.g3')
+                out_files.append(fname)
+                writer = core.G3Writer(fname)
+
+            # Initialize stuff
+            m = frame_idxs == oframe_idx
+            nsamp = np.sum(m)
+            ts = times[m]
+            data = np.zeros((self.nchans, nsamp), dtype=np.int32)
+            biases = np.zeros((len(self.bias_names), nsamp), dtype=np.int32)
+            primary = np.zeros((len(self.primary_names), nsamp), dtype=np.int64)
+            filled = np.zeros(nsamp, dtype=bool)
+            
+            # Loop through in_frames filling current out_frame
+            while True:
+                # First, write any intermediate frames like observation and wiring
+                for fr in interm_frames:
+                    if 'frame_num' in fr:
+                        del fr['frame_num']
+                    fr['frame_num'] = oframe_num  # Update this so they remain ordered
+                    oframe_num += 1
+                    writer(fr)
+
+                m = (oframe_idxs == oframe_idx) & (iframe_idxs == iframe_idx)
+                outsamps = out_offset_idxs[m]
+                insamps = in_offset_idxs[m]
+
+                # A fast way to copy data into a numpy array is to use direct
+                # slicing like ``arr_out[o0:o1] = arr_in[i0:i1]`` since it doesn't
+                # need to create a temporary copy of the array, and can just do a
+                # direct mem-map. Doing this speeds up the binding by a factor of
+                # 3-4x compared to other methods I've tried.
+                split_idxs = 1 + np.where(
```

Comment:
a comment here about what split_idxs is will be helpful for understanding the subsequent code